### DISCUSSION TITLE
Is UE in DTA necessary ?

### DISCUSSION THREAD

**Discussion by wangyinpu** [jdlph/Path4GMNS#31] [Q&A] [answered]
When i use pg.perform_network_assignment_DTALite(mode=2, column_gen_num, column_update_num)，the mode is UE + DTA, the UE process is necessary before DTA.
If I have my own agent.csv, can I just call the DNL in PATH4GMNS?
Thank you!

**Comment by jdlph** [comment, marked as answer]
UE is the prerequisite to conduct DTA for both Path4GMNS and DTALite. 

pg.perform_network_assignment_DTALite(mode=2, column_gen_num, column_update_num) is actually calling DTALite embedded in Path4GMNS. DTALite only supports an integrated run of UE + DTA. In other words, you cannot bypass the UE process and directly run DTA using your own agent.csv.

However, this specific running mode is supported by the simulation module in Path4GMNS (and will be supported in [TransOMS](https://github.com/jdlph/TransOMS). You can directly call pg.perform_simple_simulation(network) to perform DTA based on your own agent.csv. The full sample code is listed below.

```python
import path4gmns as pg

network = pg.read_network()
pg.read_zones(network)
pg.load_demand(network)

pg.load_columns(network)
pg.perform_simple_simulation(network)
print('complete dynamic simulation.\n')

print('writing agent trajectories')
pg.output_agent_trajectory(network)
```

You can even continue the UE process starting from your own agent.csv. Details can be found [here](https://path4gmns.readthedocs.io/en/latest/usecases.html#perform-path-based-ue-traffic-assignment-using-the-python-column-generation-module).

As this "issue" is more toward a question or a discussion, in the future you may want to consider [discussions](https://github.com/jdlph/Path4GMNS/discussions), which is a more proper place. If you have further questions w.r.t. your original question, you can reopen this thread.

**Comment by wangyinpu** [comment]
您好，您的邮件我已收到，谢谢。Hello, your email has been received, thank you.

**Reply by wangyinpu** [reply]
Thanks a lot, Peiheng. I found that pg.perform_simple_simulation(network) can perform DTA based on my own agent.csv. While it is still necessaary to load demands before the simulation. The demand.csv and agent.csv both determine the process of the simulation. (I tested and found that the traffic flow shared by both is loaded and output). I want to know the role and necessity of the demand.csv file in the simulation.
In addition, the trajectory.csv of pg.perform_simple_simulation(network) maybe contain two minor errors: the order of node_sequence and geometry are reversed.
Looking forward to your reply, thank you.

**Reply by jdlph** [reply]
My original answer is not quite right. You would still need to load columns (agent.csv) before conducting simulation. The code snippet has been updated accordingly.  I guess you probably mean the same thing by saying "load demands". 

The error does exist. At runtime, node path and link path are both in reverse order, which then will be reversed during the output process. In other words, load_columns() should have revered node path and link path during the loading process. We will work on the fix and include it in the next release (v0.9.2). Thank you for reporting this bug!

demand.csv is only used to set up the volume of each OD pair if you do pg.load_columns(network). This is NOT necessary as the volume of an OD pair can be easily deduced from its columns. We are planning to address it in v0.9.3.

**Reply by wangyinpu** [reply]
Thanks for the reply, looking forward to the new version.

**Reply by jdlph** [reply]
The bug w.r.t. the incorrect order of link path and geometry in trajectory.csv has been fixed under [2be00cd](https://github.com/jdlph/Path4GMNS/commit/2be00cd572ff3432e1d8c482e518c70eebad3c70) along with another fix on computing the total system travel time. 

I attached the updated package for you to test. It can be installed locally from the terminal through the following command. You will need to unzip it before installation. Please let us know your test result. Thank you!

```
python -m pip install path4gmns-0.9.2-py3-none-any.whl
```



[path4gmns-0.9.2-py3-none-any.whl.zip](https://github.com/jdlph/Path4GMNS/files/10952828/path4gmns-0.9.2-py3-none-any.whl.zip)

**Reply by wangyinpu** [reply]
The test results are good. :-）

**Reply by jdlph** [reply]
> My original answer is not quite right. You would still need to load columns (agent.csv) before conducting simulation. The code snippet has been updated accordingly. I guess you probably mean the same thing by saying "load demands".
> 
> The error does exist. At runtime, node path and link path are both in reverse order, which then will be reversed during the output process. In other words, load_columns() should have revered node path and link path during the loading process. We will work on the fix and include it in the next release (v0.9.2). Thank you for reporting this bug!
> 
> demand.csv is only used to set up the volume of each OD pair if you do pg.load_columns(network). This is NOT necessary as the volume of an OD pair can be easily deduced from its columns. We are planning to address it in v0.9.3.

The dependency of load_columns() on read_demand() has been removed under [5c355cd](https://github.com/jdlph/Path4GMNS/commit/5c355cd91c877be8a6d18011d89701dd56ea8eed).
